website/blog/2023-08-03-adding-custom-query.md

```diff
@@ -0,0 +1,136 @@
+
+If you take a look at the `ServiceInterface` interface, you’ll notice it only  contains the `register()` method. This method holds action and filter hooks for that class. Other than hooking into existing actions and filters, this method can be used to register our own filters which can be used in blocks or other classes.
+
+Basically, whenever you need to hook into actions or filters, you use a Service class for that.
+
+## Creating a new service class
+
+Creating a new service class in your project is as simple as using the following WP-CLI command:
+
+```bash
+wp boilerplate create service-example --folder=CustomQuery --file_name=CustomQuery
+```
+
+Once this new class is generated, you can add a new public method that will contain the query logic. We want our method to accept three optional arguments:
+
+- ID of the category
+- number of posts per page
+- number of the current page
+
+Having the category ID optional will allow for one more use case, and that is fetching the latest posts regardless of category. By default, WordPress sorts the posts by publish date, from newest to oldest.
+
+```php
+/**
+ * Get posts by category ID.
+ *
+ * @param int $categoryId Category ID.
+ * @param int $postsPerPage Number of posts per page.
+ * @param int $currentPage Current page number.
+ *
+ * @return WP_Query Query object.
+ */
+public function getPostsByCategory($categoryId = null, $postsPerPage = 3, $currentPage = 1): WP_Query
+{
+	$postArgs = [
+		'post_type' => 'post',
+		'cat' => $categoryId,
+		'posts_per_page' => $postsPerPage,
+		'paged' => $currentPage,
+	];
+
+	return new WP_Query($postArgs);
+}
+```
+
+To use this method, we can add it as a filter. The filter name should be added as a constant for easier maintenance. Inside the `register()` method, add the following:
+
+```php
+\add_filter(self::GET_POSTS_BY_CATEGORY, [$this, 'getPostsByCategory'], 10, 3);
+```
+
+## Using the new filter
+
+:::note
+Even if you register a filter for a method that doesn’t accept any arguments, when calling `apply_filters`, you must pass at least 1 parameter. In those cases, simply add `null` as a parameter.
+:::
+
+Here are some examples how you can use the filter:
+
+```php
+// Get 3 latest posts, regardless of category.
+$allLatestPosts = apply_filters(CustomQuery::GET_POSTS_BY_CATEGORY, null);
+
+// Get 10 latest posts from the News category. News category ID is 2.
+$latestNews = apply_filters(CustomQuery::GET_POSTS_BY_CATEGORY, 2, 10);
+
+// Get another page of News category.
+$pagedNews = apply_filters(CustomQuery::GET_POSTS_BY_CATEGORY, 2, 10, $currentPage);
+```
+
+The above example shows multiple use cases. The first two examples could be used in a simple block that displays only the selected number of the latest posts. The last example may be used in a REST route for a load more functionality or in a block with classic pagination.
+
+Now you can do a regular query loop in your block to display the posts:
```

Comment:
Even though it's there, maybe we should mention that `wp_reset_postdata()` shouldn't be forgotten.

Reply:
Yep, I agree. I'll update it 🙂